src/ftdc_tools/decode.py

```diff
@@ -0,0 +1,203 @@
+"""Package to decode FTDC data."""
+import datetime
+import io
+import zlib
+
+from collections import OrderedDict
+from collections.abc import MutableMapping
+from dataclasses import dataclass
+from typing import IO, Any, BinaryIO, Callable, Iterator, List, Tuple, Union
+
+import bson
+
+from bson.codec_options import CodecOptions
+
+# Values an FTDC value may take.
+FTDCValue = Union[datetime.datetime, int, bool, OrderedDict[str, Any]]
+
+
+# Type for a FTDC doc returned from an iter function.
+# In principle a FTDCDoc should only contain FTDCValues, but
+# we can't embed this definition in that one. (mypy doesn't
+# support cyclic definitions.)
+FTDCDoc = OrderedDict[str, Any]
+
+
+# Types for an undecoded BSON doc.
+BSONSingleValue = Union[datetime.datetime, bson.int64.Int64, bool]
+BSONValue = Union[BSONSingleValue, MutableMapping[str, Any]]
+
+
+# Path to a key in a FTDC document.
+KeyPath = Tuple[str, ...]
+
+
+@dataclass
+class MetricChunk:
+    """Dataclass representing a single chunk's metrics (one column)."""
+
+    key_path: Tuple[str, ...]
+    values: List[FTDCValue]
+    delta_constructor: Callable[..., Any]
+
+
+# We have mypy ignore CodecOptions and its uses throughout; it complains about
+# usage recommended in the pymongo bson docs.
+ftdc_bson_options = CodecOptions(document_class=OrderedDict)  # type: ignore
+
+
+def decode_iter(ftdc: bytes) -> Iterator[FTDCDoc]:
+    """Iterate over all FTDC docs in a set of bytes."""
+    for chunk in bson.decode_iter(ftdc, codec_options=ftdc_bson_options):  # type: ignore
+        for doc in _iter_chunk(chunk):  # type: ignore
+            yield doc
+
+
+def decode_file_iter(file_obj: Union[BinaryIO, IO]) -> Iterator[FTDCDoc]:
+    """Iterate over all FTDC docs in a stream."""
+    for chunk in bson.decode_file_iter(file_obj, codec_options=ftdc_bson_options):  # type: ignore
+        for doc in _iter_chunk(chunk):  # type: ignore
+            yield doc
```

Comment:
These are the main user-facing functions. I tried to have them match the pymongo equivalents.